src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.cs

```diff
@@ -46,54 +68,225 @@ internal SafeWaitHandle[] GetWaitedSafeWaitHandleArray(int requiredCapacity)
+        public string Name
+        {
+            get
+            {
+                return _name;
+            }
+            set
+            {
+                using (LockHolder.Hold(_lock))
+                {
+                    if (_name != null)
+                    {
+                        throw new InvalidOperationException(SR.InvalidOperation_WriteOnce);
+                    }
+                    _name = value;
+                    // TODO: Inform the debugger and the profiler
+                }
+            }
+        }
+
+        public ThreadPriority Priority
+        {
+            get
+            {
+                if (IsDead)
+                {
+                    throw new ThreadStateException(SR.ThreadState_Dead_Priority);
+                }
+                return _priority;
+            }
+            set
+            {
+                if ((value < ThreadPriority.Lowest) || (ThreadPriority.Highest < value))
+                {
+                    throw new ArgumentOutOfRangeException(SR.Argument_InvalidFlag);
+                }
+                if (IsDead)
+                {
+                    throw new ThreadStateException(SR.ThreadState_Dead_Priority);
+                }
+
+                // Possible race condition with starting this thread
+                using (LockHolder.Hold(_lock))
+                {
+                    if (!SetPriority(value))
+                    {
+                        throw new ThreadStateException(SR.ThreadState_SetPriorityFailed);
+                    }
+                    _priority = value;
+                }
+            }
+        }
+
+        public ThreadState ThreadState => GetThreadState();
+
+        private ThreadState GetThreadState()
+        {
+            int state = Volatile.Read(ref _threadState);
+            // If the thread is marked as alive, check if it has finished execution
+            if ((state & (int)(ThreadState.Unstarted | ThreadState.Stopped | ThreadState.Aborted)) == 0)
+            {
+                if (HasFinishedExecution())
+                {
+                    state = Volatile.Read(ref _threadState);
+                    if ((state & (int)(ThreadState.Stopped | ThreadState.Aborted)) == 0)
+                    {
+                        SetThreadStateBit(ThreadState.Stopped);
+                        state = _threadState;
+                    }
+                }
+            }
+            return (ThreadState)state & PublicThreadStateMask;
+        }
+
+        private bool GetThreadStateBit(ThreadState bit)
+        {
+            Debug.Assert((bit & ThreadState.Stopped) == 0, "ThreadState.Stopped bit may be stale; use GetThreadState instead.");
+            return (Volatile.Read(ref _threadState) & (int)bit) != 0;
+        }
+
+        private void SetThreadStateBit(ThreadState bit)
+        {
+            int oldState, newState;
+            do
+            {
+                oldState = _threadState;
```

Comment:
Sergiy's concern was about a possible compiler optimization: re-reading the _threadState field from memory (instead of using the oldState local variable) in the `while` condition of this loop:
```
            do
            {
                oldState = _threadState;
                newState = oldState | (int)bit;
            } while (Interlocked.CompareExchange(ref _threadState, newState, oldState) != oldState);
```
I decided to mark the _threadState field as volatile to make all reads of this field volatile.